runtime/array.c

```diff
@@ -414,39 +370,20 @@ static value caml_array_gather(intnat num_arrays,
     CAMLassert(pos == size);
   }
 #endif
-  else if (size <= Max_young_wosize) {
+  else if (size > Max_wosize) {
+    /* Array of values, too big. */
+    caml_invalid_argument("Array.concat");
+  }
+  else {
     /* Array of values, small enough to fit in young generation.
        We can use memcpy directly. */
     res = caml_alloc_small(size, 0);
```

Comment:
This is a bug. Should be `caml_alloc`.

Reply:
Indeed, the code was broken compared with what's in the current trunk.

General suggestion: now that there are no read barriers to account for, why not restart from runtime/array.c from the current trunk?  Apart from the atomicity concerns, the current array.c code should work, no?

Reply:
I've reverted the `array.c` changes and have made a PR: https://github.com/ocaml-multicore/ocaml-multicore/pull/822. It would benefit from a quick review.